x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/InternalCategorizationAggregation.java

```diff
@@ -0,0 +1,436 @@
+                maxMatchTokens,
+                similarityThreshold
+            );
+            // TODO: Could we do a merge sort similar to terms?
+            //  It would require us returning partial reductions sorted by key, not by doc_count
+            // First, make sure we have all the counts for equal log groups
+            Map<BucketKey, DelayedCategorizationBucket> reduced = new HashMap<>();
+            for (InternalAggregation aggregation : aggregations) {
+                InternalCategorizationAggregation categorizationAggregation = (InternalCategorizationAggregation) aggregation;
+                for (Bucket bucket : categorizationAggregation.buckets) {
+                    reduced.computeIfAbsent(bucket.key, key -> new DelayedCategorizationBucket(key, new ArrayList<>(1), 0L)).add(bucket);
+                }
+            }
+
+            reduced.values()
+                .stream()
+                .sorted(Comparator.comparing(DelayedCategorizationBucket::getDocCount).reversed())
+                .forEach(bucket ->
+                    // Parse log line takes document count into account and merging on smallest groups
+                    categorizationTokenTree.parseLogLine(hash.getIds(bucket.key.keyAsTokens()), bucket.docCount)
+                );
+            categorizationTokenTree.mergeSmallestChildren();
+            Map<BucketKey, DelayedCategorizationBucket> mergedBuckets = new HashMap<>();
+            for (DelayedCategorizationBucket delayedBucket : reduced.values()) {
+                TextCategorization group = categorizationTokenTree.parseLogLineConst(hash.getIds(delayedBucket.key.keyAsTokens()));
+                if (group == null) {
+                    throw new AggregationExecutionException(
+                        "Unexpected null categorization group for bucket [" + delayedBucket.key.asString() + "]"
+                    );
+                }
+                BytesRef[] categoryTokens = hash.getDeeps(group.getCategorization());
+
+                BucketKey key = reduceContext.isFinalReduce() ?
+                    BucketKey.withCollapsedWildcards(categoryTokens) :
+                    new BucketKey(categoryTokens);
+                mergedBuckets.computeIfAbsent(
+                        key,
+                        k -> new DelayedCategorizationBucket(k, new ArrayList<>(delayedBucket.toReduce.size()), 0L)
+                    ).add(delayedBucket);
+            }
+
+            final int size = reduceContext.isFinalReduce() == false ? mergedBuckets.size() : Math.min(requiredSize, mergedBuckets.size());
+            final PriorityQueue<Bucket> pq = new BucketCountPriorityQueue(size);
+            for (Map.Entry<BucketKey, DelayedCategorizationBucket> keyAndBuckets : mergedBuckets.entrySet()) {
+                final BucketKey key = keyAndBuckets.getKey();
+                DelayedCategorizationBucket bucket = keyAndBuckets.getValue();
+                Bucket newBucket = bucket.reduce(key, reduceContext);
```

Comment:
This reduces children even if you aren't going to keep the final buckets. It's sort of a shame. We have a thing to dance around this `TopBucketBuilder`. I don't know how complex it would be to integrate here. If its not too complex it's probably worth it.

Reply:
OTOH, if you aren't going to do inner aggregations for a while then it doesn't matter.

Reply:
I worked on trying to optimize the `reduce` based on `minDocCount`.  But, in the `else` we immediately call `-`countInnerBucket(newBucket)` and that needs to be an actual bucket (not a delayed one).

Reply:
:+1:. Let it stay as a thing we might do later. It's not a simple thing to integrate.